unified-scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,722 @@
+        transaction: &SanitizedTransaction,
+        index: usize,
+        handler_context: &HandlerContext,
+    );
+}
+
+#[derive(Debug)]
+pub struct DefaultTaskHandler;
+
+impl TaskHandler for DefaultTaskHandler {
+    fn handle(
+        result: &mut Result<()>,
+        timings: &mut ExecuteTimings,
+        bank: &Arc<Bank>,
+        transaction: &SanitizedTransaction,
+        index: usize,
+        handler_context: &HandlerContext,
+    ) {
+        // scheduler must properly prevent conflicting tx executions. thus, task handler isn't
+        // responsible for locking.
+        let batch = bank.prepare_unlocked_batch_from_single_tx(transaction);
+        let batch_with_indexes = TransactionBatchWithIndexes {
+            batch,
+            transaction_indexes: vec![index],
+        };
+
+        *result = execute_batch(
+            &batch_with_indexes,
+            bank,
+            handler_context.transaction_status_sender.as_ref(),
+            handler_context.replay_vote_sender.as_ref(),
+            timings,
+            handler_context.log_messages_bytes_limit,
+            &handler_context.prioritization_fee_cache,
+        );
+    }
+}
+
+// Currently, simplest possible implementation (i.e. single-threaded)
+// this will be replaced with more proper implementation...
+// not usable at all, especially for mainnet-beta
+#[derive(Debug)]
+pub struct PooledScheduler<TH: TaskHandler> {
+    id: SchedulerId,
+    pool: Arc<SchedulerPool<Self, TH>>,
+    context: Option<SchedulingContext>,
+    result_with_timings: Mutex<Option<ResultWithTimings>>,
+}
+
+impl<TH: TaskHandler> PooledScheduler<TH> {
+    fn do_spawn(pool: Arc<SchedulerPool<Self, TH>>, initial_context: SchedulingContext) -> Self {
+        Self {
+            id: pool.new_scheduler_id(),
+            pool,
+            context: Some(initial_context),
+            result_with_timings: Mutex::default(),
+        }
+    }
+}
+
+pub trait SpawnableScheduler<TH: TaskHandler>: InstalledScheduler {
+    fn has_context(&self) -> bool;
+
+    fn replace_context(&mut self, context: SchedulingContext);
+
+    fn spawn(pool: Arc<SchedulerPool<Self, TH>>, initial_context: SchedulingContext) -> Self
+    where
+        Self: Sized;
+}
+
+impl<TH: TaskHandler> SpawnableScheduler<TH> for PooledScheduler<TH> {
+    fn has_context(&self) -> bool {
+        self.context.is_some()
+    }
+
+    fn replace_context(&mut self, context: SchedulingContext) {
+        self.context = Some(context);
+        *self.result_with_timings.lock().expect("not poisoned") = None;
+    }
+
+    fn spawn(pool: Arc<SchedulerPool<Self, TH>>, initial_context: SchedulingContext) -> Self {
+        Self::do_spawn(pool, initial_context)
+    }
+}
+
+impl<TH: TaskHandler> InstalledScheduler for PooledScheduler<TH> {
+    fn id(&self) -> SchedulerId {
+        self.id
+    }
+
+    fn context(&self) -> &SchedulingContext {
+        self.context.as_ref().expect("active context should exist")
```

Comment:
Having a bit of trouble reasoning through the safety of this `expect`.

1. I can see when we create new banks in `BankForks` this context will always be `Some` if we have an installed scheduler pool.
2. `context` gets checked when we schedule txs, and when we return a scheduler to pool

When returning a scheduler:
1. `wait_for_termination` removes our context iff `wait_reason` is not paused
2. in `InstalledSchedulerPool::wait_for_termination` we call the scheduler `wait_for_termination`, then `return_to_pool`.
3. So if the wait reason is not paused, we call `wait_for_termination` which will take our context, then we call `return_to_pool` which will assert context is none.

Seems safe, but a little hard to track that down.

When scheduling txs:
1. We are operating on **some** bank within `BankForks`, it's difficult for me to conceptually guarantee this has **not** had the `return_to_pool` called on it.

## Actual Suggestions instead of rambling thoughts

Wishing I had caught this earlier because I'm rethinking the traits.
I wonder if we could possibly get type-safety around the context by distinctly separating the `wait_for_termination` for paused vs not paused.

It seems the scheduler is in one of two states:
1. In the pool, doing nothing, waiting for context
2. Associated with a bank/context, and being used for scheduling/execution

It seems like the pool could have some `Box<dyn InstallableScheduler>`, where `InstallableScheduler` has some fn to transition it to `Box<dyn InstalledScheduler>`.
And `InstalledScheduler` must also have a fn to transition **back** to `InstallableScheduler` before we can return to pool, via the wait for termination.

That way, we can guarantee that if we have an **installed** scheduler we always have a context.
This does make the type-system more complicated, but I'm leaning towards it being a complicated enough relationship since I'm dumb and cannot explicitly verify safety around it.
wdyt?

Reply:
thanks for suggestion. i don't like the `.unwrap()` either.. your suggestion will work for current impl. however, once multi-threaded, maintaining the type-safety is hard. also, `::context()` isn't that important for impl-wise.

so, i dcou-ed it: 8844fd58d1b

Reply:
> When scheduling txs:
> 
>     1. We are operating on **some** bank within `BankForks`, it's difficult for me to conceptually guarantee this has **not** had the `return_to_pool` called on it.

btw, `return_to_pool` consumes `self`. so, any wild schedulers outside the pool should have been `take_scheduler()`-ed without `return_to_pool` being called.

Reply:
Yeah, they will have been `take_scheduler`'ed (which guarantees they had a context at that point) at some prior point without `return_to_pool`.

What's more difficult to track is that `wait_for_termination` is what actually removes the context, and does not consume the boxed-self. If we removed the context within `return_to_pool`, that would make things more clear to me at least.
Is there a reason we cannot/should not do it there?

Reply:
> If we removed the context within return_to_pool, that would make things more clear to me at least.

thanks for explanation. so, you prefer like this, correct?: https://github.com/solana-labs/solana/pull/34300/commits/60b1bd2511a714690b0b2331e49bc3d0c72e3475

> What's more difficult to track is that wait_for_termination is what actually removes the context, and does not consume the boxed-self.

there's 2 small reasons:

1. imo, resources should be disposed as soon as it's no longer needed semantically. in this case, it's the context at the time of scheduler termination. so, I'd prefer the early dropping. also, mutating state (ie removing the context) when just _returning_ to the pool sounds a bit unnatural for me.
2. i understand concern around panic-safety in the early dropping. on the other hand, this clearly dictates that calling `wait_for_termination` _twice_ is a fatal invariant violation for the same scheduler unless paused. and ensures this won't happen on production ever.


anyway, i'm not too obsessed with this. I'm just wine with 60b1bd2511a714690b0b2331e49bc3d0c72e3475. just wanted to share some my picky opinionated codding practice. ;)

Reply:
@apfitzge good news. i did the hassle: 7334372830f. i think i did my best to apply your suggestion. and, my prev comment's semantic ramblings are upheld as well in the commit.

Reply:
I've further pushed some cleaning-up commits.

also, i had to rebase this pr with latest changes at master: https://github.com/solana-labs/solana/pull/34300/commits/d660e4203c9ca26e0ff9b140352b30e3e88d600a